pkg/sql/catalog/descs/txn.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package descs
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/config"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/database"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/lease"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/errors"
+)
+
+var errTwoVersionInvariantViolated = errors.Errorf("two version invariant violated")
+
+// Txn enables callers to run transactions with a *Collection such that all
+// retrieved immutable descriptors are properly leased and all mutable
+// descriptors are handled. The function deals with verifying the two version
+// invariant and retrying when it is violated. Callers need not worry that they
+// write mutable descriptors multiple times.
+//
+// The passed transaction is pre-emptively anchored to the system config key on
+// the system tenant.
+func Txn(
+	ctx context.Context,
+	settings *cluster.Settings,
+	leaseMgr *lease.Manager,
+	ie sqlutil.InternalExecutor,
+	db *kv.DB,
+	f func(ctx context.Context, txn *kv.Txn, descriptors *Collection) error,
```

Comment:
nit: I might prefer this argument be `descCollection` or something. `descriptors` always makes me think we're calling functions in a `descriptors` package. I don't feel strongly about this at all.